src/Dhall/TypeCheck.hs

```diff
@@ -154,11 +155,29 @@ typeWithA tpa = loop
                     then return ()
                     else Left (TypeError ctx e (AnnotMismatch a0 nf_A0 nf_A1))
             Nothing -> return ()
-        let a1 = Dhall.Core.normalize a0
-        let a2 = Dhall.Core.shift 1 (V x 0) a1
-        let b1 = Dhall.Core.subst (V x 0) a2 b0
-        let b2 = Dhall.Core.shift (-1) (V x 0) b1
-        loop ctx b2
+
+        t <- loop ctx _A1
+
+        -- The catch-all branch will actually do the correct thing for
```

Comment:
Perhaps rephrase this comment to explain that the slow path (i.e. the catch-all branch) exactly matches the standard and has to be used for right-hand-sides that are types in order to correctly support type synonyms.  For right-hand-sides that are not types we use the fast path which doesn't require type-checking the right-hand-side of the `let` expression multiple times if the bound variable is referenced repeatedly